### ISSUE TITLE
[improvement]: Allow customization of Asterisk start/stop commands

### ISSUE THREAD

**Issue by JoseGoncalves** [FreePBX/issue-tracker#461]
### FreePBX Version

FreePBX 17

### Improvement Description

If someone builds/packs it's own Asterisk, or uses some pre-built Asterisk packages (like the ones [provided by Debian](https://packages.debian.org/source/sid/asterisk)), it will most probably have a systemd service file that he can use to start/stop Asterisk.

In those situations, it will be a nice addition if someone that starts/stops Asterisk via `fwconsole start/stop/restart` commands will downwards cause the invocation of `systemctl start/stop/restart asterisk.service` commands.

I've successfully tested using `systemctl` to start/stop Asterisk in my system with these patches to FreePBX libraries:
```patch
--- /var/www/html/admin/libraries/Console/Stop.class.php.original	2024-09-23 13:46:33.000000000 +0100
+++ /var/www/html/admin/libraries/Console/Stop.class.php	2024-09-28 09:46:59.499995385 +0100
@@ -231,7 +231,7 @@
 				$sastbin = '/usr/bin/env killall safe_asterisk > /dev/null 2>&1';
 				exec($sastbin);
 			case "gracefully":
-				$astbin = '/usr/bin/env asterisk -rx "core stop ' . $method .'" &>/dev/null &';
+				$astbin = '/usr/bin/env systemctl stop asterisk.service &>/dev/null &';
 				shell_exec($astbin);
 			break;
 		}
```
```patch
--- /var/www/html/admin/libraries/Console/Start.class.php.original	2024-09-23 13:46:33.000000000 +0100
+++ /var/www/html/admin/libraries/Console/Start.class.php	2024-09-28 09:46:46.144479080 +0100
@@ -169,7 +169,7 @@
 
 	private function startAsterisk($output){
 		$output->writeln(_('Starting Asterisk...'));
-		$astbin = '/usr/bin/env safe_asterisk -U '.\FreePBX::Config()->get('AMPASTERISKUSER').' -G '.\FreePBX::Config()->get('AMPASTERISKGROUP').' > /dev/null 2>&1 &';
+		$astbin = '/usr/bin/env systemctl start asterisk.service > /dev/null 2>&1 &';
 		$process = Process::fromShellCommandline($astbin);
 		$env = $this->getDefaultEnv();
 		if(empty($env['TERM'] ?? '')) {
```

My suggestion would be to add a config variable to `/etc/freepbx.conf` that would control which commands are used to start/stop Asterisk gracefully.

By default, the old behaviour of stopping with `asterisk -rx "core stop gracefully"`  and starting with `safe_asterisk` would be used.

When that new config variable was set to a value that would cause the usage of the new start/stop commands, `systemctl` would be used to start/stop Asterisk.

**Comment by kapilgupta01** [comment]
hi @JoseGoncalves 

As you can see existing safe_asterisk is using AMPASTERISKUSER whereas you are trying to use `asterisk.service` which needs root permissions so php file which runs as `AMPASTERISKUSER ` will have permission issue.

We could however try to use option in `/etc/freepbx.conf` and use that option to decide if we want to use `asterisk.service` or existing behaviour so for advanced users like you can use `asterisk.service` whereas others can continue to use normal operation.

Feel free to raise PR with your changes to review further.

Regards,
Kapil

**Comment by JoseGoncalves** [comment]
Hi @kguptasangoma. My knowledge of PHP is limited, so I'm unable to raise a PR for an improvement as described in this issue (a variable that selects asterisk start/stop commands to use).

Nevertheless, what I can say, is that using the patch described works for me, using the following systemd service file (provided by [Debian's asterisk package](https://packages.debian.org/source/sid/asterisk)):

```ini
[Unit]
Description=Asterisk PBX
Documentation=man:asterisk(8)
Wants=network-online.target
After=network-online.target

[Service]
Type=notify
ExecStart=/usr/sbin/asterisk -g -f -p -U asterisk -G asterisk
ExecReload=/usr/sbin/asterisk -rx 'core reload'
Restart=on-failure
RestartSec=1
WorkingDirectory=/var/lib/asterisk

# The following two lines are by default set to null so as to avoid
# unnecessary console output to syslog. However, if you to, you can
# further edit /etc/asterisk/logger.conf to log output to syslog.
StandardOutput=null
StandardError=null

[Install]
WantedBy=multi-user.target
```

with an extra override file `/etc/systemd/system/asterisk.service.d/override.conf`:

```ini
[Unit]
After=mariadb.service
Wants=mariadb.service

[Service]
UMask=0002
```

**Comment by kapilgupta01** [comment]
agreed for you it may work but for general usage we need to tweak PHP code to handle both the situation , its like advanced vs normal usage.

**Comment by JoseGoncalves** [comment]
That's why I did not raise a PR for my (simple) solution, because it implies that a user already has a proper asterisk systemd service file installed, what would not be the case for everyone.

Meanwhile, can you give me an hint how to disable signature checking for those two files that I've changed?
It's annoying to have that red tampered file warning on the Dashboard...

**Comment by JoseGoncalves** [comment]
Just want to add some info on tests that I've made recently.

I found that (for a reason that I cannot grasp) the `systemctl stop asterisk.service` command does not work when invoked in `Stop.class.php`. But, the `systemctl start asterisk.service` works perfectly when invoked in `Start.class.php`.

Nevertheless, only this latest change is enough for the `freepbx.service` do not enter in conflict with my `asterisk.service`.
So, I can start/stop Asterisk independently of FreePBX using `asterisk.service` only by making the change indicated above on FreePBX's `Start.class.php` file.

**Comment by JoseGoncalves** [comment]
For anyone that uses FreePBX with an Asterisk build that already has a systemd service to manage it, I found a workaroud that does not require changes on FreePBX.

Simpy add an override to `freepbx.service` to instruct systemd to start the Asterisk prior of FreePBX. This can be achieved by running:

```bash
$ sudo systemctl edit freepbx.service
``` 

and adding into it:

```
[Unit]
After=asterisk.service
Wants=asterisk.service
```
